draft/celery-kubernetes-operator.rst

```diff
@@ -0,0 +1,273 @@
+
+Custom Controller
+-----------------
+
+`Custom controller <https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/#custom-controllers>`__ implementation to manage Celery applications(CRs). Contains the code for creation, updation, deletion and scaling handlers of the cluster.
+
+Async KOPF Handlers(Controller Implementation)
+----------------------------------------------
+
+This section contains brief overview of creation and updation handlers
+which are going to react on celery resource creation and updation
+respectively and return their status to be stored back as resource's
+status.
+
+Creation Handler
+----------------
+
+Generates deployment spec for worker and flower deployments dynamically
+based on incoming parameters specified in custom celery resource. Also
+creates the flower service to expose flower UI. Status of each children
+is sent back to be stored under parent resource status field.
+
+Additionally, it might handle the HPA object creation too if the scaling
+is to be done on native metrics(CPU and Memory utilization).
+
+Updation Handler
+----------------
+
+Updates deployment spec for worker and flower deployments(and service -
+HPA) dynamically and patch them. Status of each children is sent back to
+be stored under parent resource status field.
+
+
+Autoscaling
+===========
+
+This section covers how operator is going to handle autoscaling. We plan
+to supporting scaling based on following two metrics.
+
+Native Metrics(CPU, Memory Utilization)
+---------------------------------------
+
+If workers need to be scaled only on CPU/Memory constraints, we can
+simply create an HPA object in creation/updation handlers and it'll take
+care of scaling relevant worker deployment automatically. HPA supports
+these two metrics out of the box. For custom metrics, we need to do
+additional work.
+
+Broker Queue Length(KEDA based autoscaling)
+-------------------------------------------
+
+Queue Length based scaling needs custom metric server for an HPA to
+work. `KEDA <https://keda.sh/docs/1.5/concepts/>`__ is a wonderful
+option because it is built for the same. It provides the
+`scalers <https://keda.sh/docs/1.5/scalers/>`__ for all the popular
+brokers(RabbitMQ, Redis, Amazon SQS) supported in Celery.
+
+KEDA provides multiple ways to be deployed on a Kubernetes cluster -
+Helm, Operator Hub and Yaml. Celery Operator can package KEDA along with
+itself for distribution.
+
+
+
+Deployment Strategy
+===================
+
+Probably the best way would be distribute a Helm Chart which packages
+CRD, controller and KEDA together(More to be explored here). We'll also
+support YAML apply based deployments.
+
+Additionally, Helm approach is extensible in the sense that we can
+package additional components like preferred broker(Redis, RMQ, SQS) as
+well to start with Celery on Kubernetes out of the box without much
+efforts.
+
+
+Motivation & Rationale
+======================
```

Comment:
These are two different sections according to the template.

Reply:
Okay, I'll update the proposal to strictly reflect the template.